File: forminputjava/src/main/java/com/omarshehe/forminputjava/adapter/AutoCompleteAdapter.java
package com.omarshehe.forminputjava.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Filter;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.omarshehe.forminputjava.R;

import java.util.ArrayList;
import java.util.List;

public class AutoCompleteAdapter extends ArrayAdapter<String> {

    private Context context;
    private ItemSelectedListener mListener;
    private ArrayList<String> items, tempItems, suggestions;
    private boolean disableFilter;

    private Filter nameFilter = new Filter() {
        @Override
        public CharSequence convertResultToString(Object resultValue) {
            return ((String) resultValue);
        }

        @Override
        protected FilterResults performFiltering(CharSequence constraint) {
            if (constraint != null) {
                suggestions.clear();
                for (String item : tempItems) {
                    if (item.toLowerCase().contains(constraint.toString().toLowerCase())) {
                        suggestions.add(item);
                    }
                }
                FilterResults filterResults = new FilterResults();
                filterResults.values = suggestions;
                filterResults.count = suggestions.size();
                return filterResults;
            } else {
                return new FilterResults();
            }
        }

        @Override
        protected void publishResults(CharSequence constraint, FilterResults results) {
            if (disableFilter) {
                clear();
                for (String item : tempItems) {
                    add(item);
                    notifyDataSetChanged();
                }
            } else {
                List<String> filterList = (ArrayList<String>) results.values;
                if (results.count > 0) {
                    clear();
                    for (String item : filterList) {
                        add(item);
                        notifyDataSetChanged();
                    }
                }
            }
        }
    };

    public AutoCompleteAdapter(Context context, int resource, int textViewResourceId, ArrayList<String> items, ItemSelectedListener listener) {
        super(context, resource, textViewResourceId, items);
        this.items = items;
        mListener = listener;
        this.context = context;
        suggestions = new ArrayList<>();
        tempItems = new ArrayList<>(items); // this makes the difference.
    }

    public void disableFilter(boolean disableFilter) {
        this.disableFilter = disableFilter;
    }


    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        View view = convertView;
        if (convertView == null) {
            LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            view = inflater.inflate(R.layout.listitem_autocomplete, parent, false);
        }
        final String item = items.get(position);
        if (item != null) {
            TextView lblName = view.findViewById(R.id.txtAutocomplete);
            if (lblName != null)
                lblName.setText(item);

            ((LinearLayout) view.findViewById(R.id.layParent)).setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    mListener.onItemSelected(item);
                }
            });
        }


        return view;
    }

    @Override
    public Filter getFilter() {
        return nameFilter;
    }

    public interface ItemSelectedListener {
        public void onItemSelected(String item);
    }


}

File: forminputkotlin/src/main/java/com/omarshehe/forminputkotlin/adapter/AutoCompleteAdapter.kt
package com.omarshehe.forminputkotlin.adapter

import android.annotation.SuppressLint
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.Filter
import android.widget.TextView
import com.omarshehe.forminputkotlin.R
import com.omarshehe.forminputkotlin.interfaces.ItemSelectedListener
import com.omarshehe.forminputkotlin.utils.isNotTrue
import com.omarshehe.forminputkotlin.utils.isTrue
import java.util.*

class AutoCompleteAdapter(
    context: Context,
    resource: Int,
    private val items: List<String>,
    private val mListener: ItemSelectedListener
) : ArrayAdapter<String>(context, resource, items) {

    private val itemsAll: MutableList<String> = items.toMutableList()
    private val suggestions: ArrayList<String> = ArrayList()
    private var disableFilter: Boolean = false

    fun disableFilter(disableFilter: Boolean) {
        this.disableFilter = disableFilter
    }

    override fun getFilter(): Filter {
        return nameFilter
    }

    private val nameFilter: Filter = object : Filter() {
        override fun convertResultToString(resultValue: Any): String {
            return resultValue as String
        }

        @SuppressLint("DefaultLocale")
        override fun performFiltering(constraint: CharSequence): FilterResults {
            return run {
                suggestions.clear()
                itemsAll.forEach{
                    if (it.toLowerCase().contains(constraint.toString().toLowerCase())) {
                        suggestions.add(it)
                    }
                }
                val filterResults = FilterResults()
                filterResults.values = suggestions
                filterResults.count = suggestions.size
                filterResults
            }
        }

        override fun publishResults(constraint: CharSequence, results: FilterResults) {
            disableFilter.isTrue{
                clear()
                itemsAll.forEach{
                    add(it)
                    notifyDataSetChanged()
                }
            }.isNotTrue {
                val filteredList = results.values as MutableList<*>
                if (results.count > 0) {
                    clear()
                    filteredList.forEach { it as String
                        add(it)
                    }
                    notifyDataSetChanged()
                }
            }
        }
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var view = convertView
        if (view == null) {
            val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
            view = inflater.inflate(R.layout.listitem_autocomplete, parent, false)
        }
        val item = items[position]
        view!!.run {
            findViewById<TextView>(R.id.tvView).text = item
            findViewById<View>(R.id.layParent).setOnClickListener { mListener.onItemSelected(item) }
        }
        return view
    }
}